### PR TITLE
Fix problems discovered by @phpstan

### PR DESCRIPTION
**Pull request by szepeviktor**
The only hard one was `if ( ! count($args) > 3)`

### CODE REVIEW

**Comment by szepeviktor** [comment]
if you disable loading of `Patchwork.php` you can continue with
`vendor/bin/phpstan analyse -l 4`
be aware of https://github.com/phpstan/phpstan/issues/1482

**Comment by szepeviktor** [comment]
Have I made a mistake? Test are red 🔴

File: src/Container.php [hunk 1]
```diff
@@ -19,7 +19,7 @@ final class Container
 {
 
     /**
-     * @var Container
+     * @var self|null
```
Comment (gmazzap) [inline src/Container.php]: This class is final and extends nothing, so and instance of it can only be an instance of it. In such cases, I prefer to be explicit in declaring the type, so I prefer `Container` over `self`

File: src/Container.php [hunk 2]
```diff
@@ -31,11 +31,11 @@ final class Container
     /**
      * Static instance lookup.
      *
-     * @return static
+     * @return self
```
Comment (gmazzap) [inline src/Container.php]: Because of what said above, I would change to `Container` not `self`.

File: src/Container.php [hunk 3]
```diff
@@ -31,11 +31,11 @@ final class Container
...
      */
     public static function instance()
     {
-        if (!self::$instance) {
+        if (! self::$instance instanceof self) {
```
Comment (gmazzap) [inline src/Container.php]: I understand this makes phpstan happy, but in PHP does not exists an empty/false object, any objects is casted to true, even an empty `stdClass`. So if a type is `null|object` checking for its boolean value is perfectly fine. This makes it more verbose (and at micro level possibly less performant) for no reason. Actually, enforce this opinionated code style is the main reason I don't like php stan: I want static analisys to find bugs, not find code style issue.

File: src/Expectation/EscapeHelper.php [hunk 4]
```diff
@@ -21,7 +21,7 @@ class EscapeHelper
 {
 
     /**
-     * @param $text
+     * @param string $text
```
Comment (gmazzap) [inline src/Expectation/EscapeHelper.php]: 👍

File: src/Expectation/Expectation.php [hunk 5]
```diff
@@ -164,7 +164,7 @@ public function __call($name, array $arguments = [])
     }
 
     /**
-     * @return \Mockery\Expectation|\Mockery\CompositeExpectation
+     * @return \Mockery\Expectation|\Mockery\ExpectationInterface
```
Comment (gmazzap) [inline src/Expectation/Expectation.php]: This is "correct", however, because of how \Mockery is written `ExpectationInterface` is missing many methods that `CompositeExpectation` has. And I can't use `CompositeExpectation` in the constructor, because `CompositeExpectation` is created in a "fluent" interface.
This means that if I use `ExpectationInterface` the IDE will not suggest a lot of methods that are actually there.

These are the consequences of using fluent methods and magic methods. In this case I have to choose between a correct doc bloc and poor dev experience (no autocomplete) and wrong doc bloc and better dev experience... and honestly I prefer the latest.

File: src/Expectation/FunctionStub.php [hunk 6]
```diff
@@ -185,8 +185,8 @@ public function echoArg($arg_num = 1)
     }
 
     /**
-     * @param  int $arg_num
-     * @return bool
+     * @param  mixed $arg_num
+     * @return int
```
Comment (gmazzap) [inline src/Expectation/FunctionStub.php]: 👍

File: src/Expectation/FunctionStub.php [hunk 7]
```diff
@@ -210,7 +210,7 @@ private function assertRedefined($function_name)
     }
 
     /**
-     * @param        $value
+     * @param mixed  $value
```
Comment (gmazzap) [inline src/Expectation/FunctionStub.php]: 👍

File: src/Hook/HookStorage.php [hunk 8]
```diff
@@ -122,8 +122,8 @@ public function isHookAdded($type, $hook, $function = null)
     }
 
     /**
-     * @param $type
-     * @param $hook
+     * @param string $type
+     * @param string $hook
```
Comment (gmazzap) [inline src/Hook/HookStorage.php]: 👍

File: src/Hook/HookStorage.php [hunk 9]
```diff
@@ -211,7 +211,7 @@ private function parseArgsToAdd(array $args, $key, $type)
             throw Exception\InvalidHookArgument::forEmptyArguments($key, $type);
         }
 
-        if ( ! count($args) > 3) {
+        if ( count($args) > 3) {
```
Comment (gmazzap) [inline src/Hook/HookStorage.php]: 👍 thanks for this.

File: src/Name/CallbackStringForm.php [hunk 10]
```diff
@@ -105,7 +105,7 @@ private function parseCallback($callback)
 
     /**
      * @param string $callback
-     * @return bool|string
```
Comment (gmazzap) [inline src/Name/CallbackStringForm.php]: 👍

File: src/Name/CallbackStringForm.php [hunk 11]
```diff
@@ -57,7 +57,7 @@ public function __toString()
     }
 
     /**
-     * @param $callback
+     * @param mixed $callback
```
Comment (gmazzap) [inline src/Name/CallbackStringForm.php]: 👍

File: src/Name/Exception/InvalidCallable.php [hunk 12]
```diff
@@ -19,7 +19,7 @@ class InvalidCallable extends Exception
 {
 
     /**
-     * @param $callback
+     * @param mixed $callback
```
Comment (gmazzap) [inline src/Name/Exception/InvalidCallable.php]: 👍

File: src/Name/Exception/InvalidClosureParam.php [hunk 13]
```diff
@@ -24,7 +24,7 @@ class InvalidClosureParam extends Exception
     const CODE_MULTIPLE_VARIADIC = 3;
 
     /**
-     * @param $name
+     * @param string $name
```
Comment (gmazzap) [inline src/Name/Exception/InvalidClosureParam.php]: 👍

File: src/Name/Exception/InvalidClosureParam.php [hunk 14]
```diff
@@ -36,8 +36,8 @@ public static function forInvalidName($name)
     }
 
     /**
-     * @param $type
-     * @param $name
+     * @param string $type
+     * @param string $name
```
Comment (gmazzap) [inline src/Name/Exception/InvalidClosureParam.php]: 👍

File: src/Name/Exception/InvalidClosureParam.php [hunk 15]
```diff
@@ -49,7 +49,7 @@ public static function forInvalidType($type, $name)
     }
 
     /**
-     * @param $name
+     * @param string $name
```
Comment (gmazzap) [inline src/Name/Exception/InvalidClosureParam.php]: 👍

File: src/Name/Exception/InvalidName.php [hunk 16]
```diff
@@ -33,7 +33,7 @@ public static function forFunction($function)
     }
 
     /**
-     * @param $class
+     * @param string $class
```
Comment (gmazzap) [inline src/Name/Exception/InvalidName.php]: 👍

File: src/Name/Exception/InvalidName.php [hunk 17]
```diff
@@ -42,7 +42,7 @@ public static function forClass($class)
     }
 
     /**
-     * @param $function
+     * @param string $function
```
Comment (gmazzap) [inline src/Name/Exception/InvalidName.php]: 👍

File: src/Name/Exception/InvalidName.php [hunk 18]
```diff
@@ -51,7 +51,7 @@ public static function forMethod($function)
     }
 
     /**
-     * @param string $thing
+     * @param mixed $thing
```
Comment (gmazzap) [inline src/Name/Exception/InvalidName.php]: 👍

File: src/Name/FunctionName.php [hunk 19]
```diff
@@ -76,7 +76,7 @@ public function equals(FunctionName $name)
      * Checks the name of a function and throw an exception if is not valid.
      * When name is valid returns an array of the name itself and its namespace parts.
      *
-     * @param  string $function_name
+     * @param  string|null $function_name
```
Comment (gmazzap) [inline src/Name/FunctionName.php]: "In theory" this should always be a string. Unfortunately not being PHP 7+, I can't use scalar type declaration in the constructor, so I can say that I "expect" a string, but can't be sure.

So if declare doc bloc based on what I expect this should be just `string`, if I declare based on what it could actually be, then it should be `mixed` because people can pass what they want in constructor.

So `string|null` is quite arbitrary...

**Comment by gmazzap** [comment]
Thanks a lot for your work.

As you already said you were able to find a bug.

Most of the doc bloc fixes were also were also very fine, and I definetevely want ot merge those.
For some other I commented why I would like to keep what I have.

That said, I also want ot write a unit test for the bug you discovered.

Because I want to do a release today (while I have some time) I think it is faster if I refuse this PR and do myself a commit with most of the changes you suggested, but will make sure to credit you where due.

Thanks a lot for your work, I really appreaciate it.

**Comment by szepeviktor** [comment]
I understand you points.
When I (_secretly_) develop code daily I use PHPStan to check my work - I'm not an IDE user.

File: src/Container.php (hunk 3 above)
Comment (szepeviktor) [reply src/Container.php]: putting a non-boolean value in if() stop you from writing robust code

Comment (jrfnl) [reply src/Container.php]: > in PHP does not exists an empty/false object, any objects is casted to true

Not necessarily true.... See: https://phpcheatsheets.com/test/boolean_tests/ (check the last two columns and the last few rows)

Comment (gmazzap) [reply src/Container.php]: > Not necessarily true...
The only case in which an object implicitly cast to boolean returns false is the object obtained via `SplBool(false)`. And `SplBool` is a class that comes from an extension not bundled with PHP (see https://www.php.net/manual/en/spl-types.installation.php). Meaning, that if we consider only classes in PHP _core_, the statement is always true.

> putting a non-boolean value in if() stops you from writing robust code
@szepeviktor that's very opinionated, and I kindly disagree.
As said above, if you are sure that a variable is `ClassName|null` (and that `ClassName` is not `SplBool`, to make @jrfnl happy :)), then `if ($var)` is identical to if ($var instanceof ClassName). And saying that two statements with 100% identical results have different "robustness" is nonsensical to me: they can be more/less readable, they can be more/less performant, but they can't be less/more robust, as the robustness depends on how they behave on different inputs and conditions, and if they always behave the same, I don't see how they can be differently robust.
Moreover, in PHP implicit casting has its rules that you can leverage.
For example if a variable is of type `array|null` then `if ($var)` is equivalent to `if ($var !== null && $var !== [])`, and if that is what you want to test for, then `if ($var)` is much more readable and much less error prone (it's easy to mistype `!=` for `!==`).
So in that specific case, IMO, not putting a boolean in `if` is much better than doing it.
In short, the implicit boolean casting of variables is a _tool_ in PHP. And just like any tool might be used for good.
